server/src/main/java/org/elasticsearch/index/mapper/MapperMergeContext.java

```diff
@@ -43,10 +49,113 @@ private MapperMergeContext(MapperBuilderContext mapperBuilderContext) {
      * @return a new {@link MapperMergeContext} with this context as its parent
      */
     public MapperMergeContext createChildContext(String name) {
-        return new MapperMergeContext(mapperBuilderContext.createChildContext(name));
+        return createChildContext(mapperBuilderContext.createChildContext(name));
+    }
+
+    /**
+     * Creates a new {@link MapperMergeContext} with a given {@link MapperBuilderContext}
+     * @param childContext the child {@link MapperBuilderContext}
+     * @return a new {@link MapperMergeContext}, wrapping the provided {@link MapperBuilderContext}
+     */
+    public MapperMergeContext createChildContext(MapperBuilderContext childContext) {
+        return new MapperMergeContext(childContext, remainingFieldsBudget);
     }
 
     MapperBuilderContext getMapperBuilderContext() {
         return mapperBuilderContext;
     }
+
+    void removeRuntimeField(Map<String, RuntimeField> runtimeFields, String name) {
```

Comment:
Now that I think of this more: removing a runtime field can only be done via static declaration of mappings, hence I don't think it should be accounted for here? you need to set the field to null in a put mapping call... removing this would spare us the need for the increment method in the budget.

I was also thinking: once we reach the limit, we are never going to go backwards, meaning the budget can effectively only ever be decremented (if we agree that remove should go away)? In that case, we could consider swapping the budget to a no-op that always returns false, instead of checking of nothing: we already know we reached the limit? Does this make sense?

Reply:
Nevermind, what I suggested is probably overkill.